File: solutions/fortran/bob/2/bob.f90
module bob
  implicit none
contains

  character(len=34) function hey(statement)
    character(*), intent(in) :: statement
    character(len=1) :: c

    character(len=34), parameter :: fine = "Fine. Be that way!"
    character(len=34), parameter :: calm = "Calm down, I know what I'm doing!"
    character(len=34), parameter :: chill = "Whoa, chill out!"
    character(len=34), parameter :: sure = "Sure."
    character(len=34), parameter :: whatever = "Whatever."
    
    logical :: has_lowercase, has_uppercase, is_question, is_silent
    integer :: i, statement_len

    has_lowercase = .false.
    has_uppercase = .false.
    is_silent = .true.
    statement_len = len(trim(statement))
    
    do i = 1,len(statement)
      if (statement(i:i) /= ' ') then
        is_silent = .false.
      end if
      if (statement(i:i) >= 'a' .and. statement(i:i) <= 'z') then
        has_lowercase = .true.
      else if (statement(i:i) >= 'A' .and. statement(i:i) <= 'Z') then
        has_uppercase = .true.
      end if
    end do

    ! Check for terminating question mark
    is_question = statement(statement_len:statement_len) == '?'

    if (is_silent) then
      hey = fine
    else if (.not. has_lowercase .and. has_uppercase .and. is_question) then
      hey = calm
    else if (.not. has_lowercase .and. has_uppercase) then
      hey = chill
    else if (is_question) then
      hey = sure
    else
      hey = whatever
    end if

  end function hey

end module bob

File: solutions/cpp/bob/2/bob.cpp
#include "bob.h"

#include <cctype>
#include <map>

namespace bob
{
    enum class response_type
    {
        question,
        forceful_question,
        statement,
        forceful_statement,
        silence
    };

    using response_t = enum class response_type;

    std::map<response_t, std::string> responses{
        {response_type::question, "Sure."},
        {response_type::forceful_question, "Calm down, I know what I'm doing!"},
        {response_type::statement, "Whatever."},
        {response_type::forceful_statement, "Whoa, chill out!"},
        {response_type::silence, "Fine. Be that way!"}
    };

    response_t classify(const std::string &statement)
    {
        bool has_char = false;
        bool has_alpha = false;
        bool is_forceful = true;
        bool is_question = false;

        for (auto it = statement.rbegin(); it < statement.rend(); it++)
        {
            if (!std::isspace(*it))
            {
                /* A question is defined as the trailing non whitespace being '?' */
                if (!has_char && *it == '?')
                {
                    is_question = true;
                }

                has_char = true;

                if (std::islower(*it))
                {
                    has_alpha = true;
                    is_forceful = false;
                }
                else if (std::isalpha(*it))
                {
                    has_alpha = true;
                }
            }
        }

        /* 
         * Because assume forceful and disprove by checking for lowercase, 
         * if we've found no alphabetic characters at all, set to false  
         */
        if (!has_alpha)
        {
            is_forceful = false;
        }

        if (!has_char)
            return response_type::silence;

        if (is_forceful && is_question)
            return response_type::forceful_question;

        if (is_question)
            return response_type::question;

        if (is_forceful)
            return response_type::forceful_statement;

        return response_type::statement;
    }

    std::string hey(const std::string &statement)
    {
        return responses.at(classify(statement));
    }

} // namespace bob
